File: src/main/java/waterpower/client/render/item/IItemIconContainerProvider.java
package waterpower.client.render.item;

import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import waterpower.client.render.IIconContainer;

public interface IItemIconContainerProvider extends IItemIconProvider {

    @Override
	@SideOnly(Side.CLIENT)
    default TextureAtlasSprite getIcon(ItemStack stack, int pass) {
        IIconContainer iconContainer = getIconContainer(stack);
        if(iconContainer != null) {
            switch (pass) {
                case 0:
                    return iconContainer.getIcon();
                case 1:
                    return iconContainer.getOverlayIcon();
            }
        }
        return null;
    }

    @Override
	@SideOnly(Side.CLIENT)
    default int getRenderPasses(ItemStack stack) {
        return 1;
    }

	@SideOnly(Side.CLIENT)
    IIconContainer getIconContainer(ItemStack itemStack);

}

File: src/main/kotlin/waterpower/client/render/item/IItemIconContainerProvider.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.client.render.item

import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.item.ItemStack
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import waterpower.client.render.IIconContainer

interface IItemIconContainerProvider : IItemIconProvider {

    @SideOnly(Side.CLIENT)
    fun getIconContainer(stack: ItemStack): IIconContainer?

    @SideOnly(Side.CLIENT)
    override fun getIcon(stack: ItemStack, layer: Int): TextureAtlasSprite? {
        val iconContainer = getIconContainer(stack)
        return when (layer) {
            0 -> iconContainer?.getIcon()
            1 -> iconContainer?.getOverlayIcon()
            else -> null
        }
    }

    @SideOnly(Side.CLIENT)
    override fun getRenderLayer(stack: ItemStack) = 1
}